pages/blog/_posts/debouncing-in-queuing-systems-optimizing-efficiency-in-async-workflows.mdx

```diff
@@ -0,0 +1,171 @@
+
+```sql
+CREATE UNIQUE INDEX debounce_key_status 
+ON debounce_job_queue(debounce_key, status)
+WHERE status = 'QUEUED';
+```
+
+Creating a new job in this queue is a relatively straightforward insert, but we add a few things that help us implement debounce:
+
+1. Create a deterministic `debounce_key` using a hash of the job's type and payload JSON data. Alternatively, this could be any unique piece of data like a user id.
+2. Allow the calling function to specify a delay when setting the `scheduled_at`.
+3. Handle conflicts with existing queued items with an `UPDATE` to the existing item's `scheduled_at` with the specific delay. **This is where the debouncing happens.**
+
+We'll use the `@vercel/postgres` library's simple `sql` here, but your favorite library will do just fine.
+
+```jsx
+import { crypto } from 'node:crypto'
+import { sql } from '@vercel/postgres'
+
+async function enqueue(jobType, payload, delay = 0) {
+	const json = JSON.stringify(payload)
+	const debounceKey = crypto.createHash('sha256')
+		.update(jobType + json).digest('hex')
+	return await sql`
+		INSERT INTO debounce_job_queue (
+			job_type, status, payload, scheduled_at, debounce_key
+		)
+		VALUES (
+			'${jobType}', 'QUEUED', '${json}',
+			now() + INTERVAL '${delay} seconds', ${debounceKey}
+		)
+		-- If the item is already QUEUED, update the scheduled at w/ the delay
+		ON CONFLICT (debounce_key, status) WHERE status = 'QUEUED'
+		DO UPDATE SET scheduled_at = now() + INTERVAL '${delay} seconds';
+	`
+}
+```
+
+The code above may appear complex at first, but the magic happens in the `ON CONFLICT` statement. Our unique index ensures that two identical jobs cannot be `'QUEUED'` at the same time. When the query attempts to insert a job and there is an identical job that is `QUEUED`, the conflict statement is executed and the existing job's scheduled at is updated with our delay to push it back.
+
+To query for available jobs, we'll also now use our `scheduled_at` timestamp to ensure that our function only fetches jobs that are ready to be processed. When jobs are debounced and the `scheduled_at` is pushed into the future, that prevents our query from fetching that job.
+
+In our query, we'll also update the `status` to `PROCESSING` to mark that this job should not be picked up by any other worker and that it shouldn't be debounced again as it's being executed. In our code, we'll omit the worker polling logic and focus on the query. We'll use `SKIP LOCKED` here to prevent only query for rows that are locked due to being updated elsewhere.
+
+```jsx
+async function getQueueItem(jobType) {
+	return await sql`
+		UPDATE debounce_job_queue
+		SET status = 'PROCESSING'
+		WHERE id = (
+		  SELECT id FROM debounce_job_queue
+		  WHERE status = 'QUEUED' and scheduled_at <= now() + INTERVAL '2 minute'
+		  ORDER BY scheduled_at ASC
+		  FOR UPDATE SKIP LOCKED
+		  LIMIT 1
+		)
+		RETURNING id, payload;
+	`
+}
+```
```

Comment:
```suggestion
```js
async function getQueueItem(jobType) {
  return await sql`
    UPDATE debounce_job_queue
    SET status = 'PROCESSING'
    WHERE id = (
      SELECT id FROM debounce_job_queue
      WHERE status = 'QUEUED' and scheduled_at <= now() + INTERVAL '2 minute'
      ORDER BY scheduled_at ASC
      FOR UPDATE SKIP LOCKED
      LIMIT 1
    )
    RETURNING id, payload;
  `;
}
```
```